### PR TITLE
Compat writemime, strings, os, startup file with some tests

### PR DESCRIPTION
**Pull request by wookay**
deprecated `writemime` to `show`.
thanks.

### CODE REVIEW

**Comment by tkelman** [comment]
Would be good to update the minimum required version of Compat in `REQUIRE` to the first tag that implements these features.

**Comment by wookay** [comment]
actually, it's not working for Julia 0.4 and 0.5
I'm trying to investigate both of IJulia.jl and Interact.jl.
then will commit again with proper REQUIRE.

thanks.

File: test/comm.jl [hunk 1]
```diff
@@ -0,0 +1,16 @@
+if VERSION >= v"0.5-"
+    using Base.Test
+else
+    using BaseTestNext
+    const Test = BaseTestNext
+end
+
+import IJulia: init, Comm
+
+profile = normpath(Pkg.dir("IJulia"), "test", "profile.json")
```
Comment (tkelman) [inline test/comm.jl]: use `dirname(@__FILE__)` rather than `Pkg.dir`

Comment (wookay) [inline test/comm.jl]: applied quickly, thanks :)

File: REQUIRE [hunk 2]
```diff
@@ -1,6 +1,6 @@
 julia 0.3
-Nettle 0.2
-JSON 0.5
-ZMQ 0.3
-Compat 0.7.15
-Conda 0.1.5
+Compat 0.8.0
+Nettle 0.2.2
+JSON 0.5.2
+ZMQ 0.3.2
```
Comment (stevengj) [inline REQUIRE]: Why did you need to increment all of these version numbers?

File: deps/build.jl [hunk 3]
```diff
@@ -62,7 +62,7 @@ juliakspec = abspath(spec_name)
 binary_name = @windows? "julia.exe":"julia"
 kernelcmd_array = String[joinpath(JULIA_HOME,("$binary_name")), "-i"]
 ijulia_dir = get(ENV, "IJULIA_DIR", Pkg.dir("IJulia")) # support non-Pkg IJulia installs
-append!(kernelcmd_array, ["-F", joinpath(ijulia_dir,"src","kernel.jl"), "{connection_file}"])
+append!(kernelcmd_array, ["--startup-file=yes", joinpath(ijulia_dir,"src","kernel.jl"), "{connection_file}"])
```
Comment (stevengj) [inline deps/build.jl]: What is the reason for this change?

File: src/comm_manager.jl [hunk 4]
```diff
@@ -54,10 +54,10 @@ function comm_info_request(sock, msg)
         comms
     end
 
-    _comms = map(reply) do x
-        x[1] => @compat Dict(:target_name, comm_target(x[2]))
+    _comms = map(collect(reply)) do x
```
Comment (stevengj) [inline src/comm_manager.jl]: Why is `collect` needed here?

File: src/msg.jl [hunk 5]
```diff
@@ -2,6 +2,18 @@ import Base.show
 
 export Msg, msg_pub, msg_reply, send_status, send_ipython
 
+# Julia commit 77631ce7f615059c7021756ef539d36964af13dc
+if VERSION >= v"0.5-dev+4194"
+    import Base.unsafe_string
+else
+    const unsafe_string = Base.bytestring
+end
```
Comment (stevengj) [inline src/msg.jl]: This shouldn't be necessary.   Compat 0.8 defines an `unsafe_string` method on 0.4.

File: src/msg.jl [hunk 6]
```diff
@@ -2,6 +2,18 @@ import Base.show
...
+
+if length(methods(unsafe_string, (ZMQ.Message,))) == 0
+    unsafe_string(zmsg::ZMQ.Message) = unsafe_string(pointer(zmsg), length(zmsg))
+end
```
Comment (stevengj) [inline src/msg.jl]: This should be patched in ZMQ.jl, not here.  Until ZMQ is patched, we can live with the deprecation message.

File: src/stdio.jl [hunk 7]
```diff
@@ -176,7 +178,7 @@ end
 function oslibuv_flush()
     #refs: https://github.com/JuliaLang/IJulia.jl/issues/347#issuecomment-144505862
     #      https://github.com/JuliaLang/IJulia.jl/issues/347#issuecomment-144605024
-    @windows_only ccall(:SwitchToThread, stdcall, Void, ())
+    Compat.is_windows() && ccall(:SwitchToThread, stdcall, Void, ())
```
Comment (stevengj) [inline src/stdio.jl]: No need for `Compat.`

File: test/msg.jl [hunk 8]
```diff
@@ -0,0 +1,21 @@
+if VERSION >= v"0.5-"
+    using Base.Test
+else
+    using BaseTestNext
+    const Test = BaseTestNext
+end
+
```
Comment (stevengj) [inline test/msg.jl]: These `using` statements don't need to be here since they are in `runtests.jl` ... I don't think we need to worry about the case where people want to runt the different tests individually.

File: test/comm.jl [hunk 9]
```diff
@@ -0,0 +1,17 @@
+if VERSION >= v"0.5-"
+    using Base.Test
+else
+    using BaseTestNext
+    const Test = BaseTestNext
+end
+
+import IJulia: init, Comm, comm_target
```
Comment (stevengj) [inline test/comm.jl]: No need to import `IJulia.init` if you are going to call it as a fully qualified name anyway.

File: deps/build.jl (hunk 3 above)
Comment (tkelman) [inline deps/build.jl]: it's deprecated on 0.5

File: REQUIRE (hunk 2 above)
Comment (wookay) [inline REQUIRE]: for Compat 0.8.0
- `bytestring(::Ptr, ...)` has been replaced with `unsafe_string`

for other packages, keeping the version numbers.

Comment (tkelman) [inline REQUIRE]: 0.5.2 isn't a good tag of JSON, it was accidentally breaking and later 0.5.x tags were actually earlier code

Comment (stevengj) [inline REQUIRE]: You should just need to bump the Compat version number.

File: REQUIRE [hunk 10]
```diff
@@ -1,6 +1,6 @@
 julia 0.3
 Nettle 0.2
-JSON 0.5
+JSON
```
Comment (stevengj) [inline REQUIRE]: Don't we require JSON 0.5 or later?  Why remove the version number?

File: REQUIRE [hunk 11]
```diff
@@ -1,6 +1,6 @@
...
 ZMQ 0.3
-Compat 0.7.15
-Conda 0.1.5
+Compat 0.8.0
+Conda 0.1.8
```
Comment (stevengj) [inline REQUIRE]: What do we require from Conda 0.1.8?

File: src/execute_request.jl [hunk 12]
```diff
@@ -72,7 +73,7 @@ end
 
 # return the content of a pyerr message for exception e
 function error_content(e, bt=catch_backtrace(); backtrace_top::Symbol=:execute_request_0x535c5df2, msg::AbstractString="")
-    tb = map(utf8, @compat(split(sprint(show_bt,
+    tb = map(String, @compat(split(sprint(show_bt,
```
Comment (stevengj) [inline src/execute_request.jl]: `map(Compat.UTF8String, ...` ?

File: src/execute_request.jl [hunk 13]
```diff
@@ -133,7 +134,7 @@ function helpcode(code::AbstractString)
         @compat if !haskey(Docs.keywords, Symbol(code_))
             return "Base.Docs.@repl $code_"
         elseif VERSION < v"0.5.0-dev+3831"
-            return "eval(:(Base.Docs.@repl \$(symbol(\"$code_\"))))"
+            return "eval(:(Base.Docs.@repl \$(Symbol(\"$code_\"))))"
```
Comment (stevengj) [inline src/execute_request.jl]: needs `@compat Symbol` for 0.3?

File: src/stdio.jl [hunk 14]
```diff
@@ -176,7 +176,7 @@ end
 function oslibuv_flush()
     #refs: https://github.com/JuliaLang/IJulia.jl/issues/347#issuecomment-144505862
     #      https://github.com/JuliaLang/IJulia.jl/issues/347#issuecomment-144605024
-    @windows_only ccall(:SwitchToThread, stdcall, Void, ())
+    is_windows() && ccall(:SwitchToThread, stdcall, Void, ())
```
Comment (stevengj) [inline src/stdio.jl]: `@static is_windows() && ...` to avoid the runtime branch

File: test/REQUIRE [hunk 15]
```diff
@@ -0,0 +1,2 @@
+julia 0.3
+Compat 0.8.0
```
Comment (stevengj) [inline test/REQUIRE]: Why is this `REQUIRE` file necessary?  Aren't these requirements already implied by the top-level `REQUIRE`?   (I thought you only need a `test/REQUIRE` if your tests require additional packages not needed for the package itself.)

File: src/execute_request.jl (hunk 13 above)
Comment (yuyichao) [inline src/execute_request.jl]: Also note that this should be needed. The branch that uses `Symbol` is below

Comment (yuyichao) [inline src/execute_request.jl]: It is written this way since there's already a branch and I was too lazy to figure out an order that `eval`, `@repl` and `@compat` are all happy about and that the `@compat` is evaluated in the right module.

File: REQUIRE (hunk 2 above)
Comment (wookay) [inline REQUIRE]: I have come to realize that the compatibility of the dependent packages.
I was not mind carefully enough to that. I'll take care.
I found these lower bounds within the commit.
Compat.jl : v0.7.20 for is_apple, is_windows
ZMQ.jl :  v0.3.1 to fix FDWatcher error during test

File: src/execute_request.jl (hunk 13 above)
Comment (wookay) [inline src/execute_request.jl]: running in the jupyter notebook with @compat,
getting an error because that has not yet imported properly for eval(expr).
I have revert back to the original code. added a related test.

**Comment by stevengj** [comment]
Some of the test failures seem spurious...

**Comment by wookay** [comment]
I'm trying to resolve these errors.

* error from appveyor (0.3, 0.5)
```
===============================[ ERROR: Nettle ]================================
failed process: Process(`7z x -y 'C:\Users\appveyor\.julia\v0.3\WinRPM\cache\1\noarch%2Fmingw32-libgcc_s_sjlj1-6.1.0-4.2.noarch.cpio' '-oC:\Users\appveyor\.julia\v0.3\WinRPM\deps'`, ProcessExited(2)) [2]
while loading C:\Users\appveyor\.julia\v0.3\Nettle\deps\build.jl, in expression starting on line 39
```

* error from travis (0.5 osx)
```
in _precompile_ at /Users/travis/.julia
jl_uv_writecb() ERROR: operation canceled ECANCELED
jl_uv_writecb() ERROR: operation canceled ECANCELED
...
```

**Comment by stevengj** [comment]
The AppVeyor failure seems like an unrelated glitch in their server; might go away if you just restart the check with a forced push.

**Comment by wookay** [comment]
on OSX, Julia 0.5 with option `inline=no` has segmentation fault
during call `start_heartbeat` (IJulia/src/heartbeat.jl)

```
~/.julia/v0.5/IJulia/test $ julia --inline=no comm.jl
[1]    51285 segmentation fault
~/.julia/v0.5/IJulia/test $ julia --inline=yes comm.jl
```

Running `Pkg.test("IJulia", coverage=true)`, the argument coverage makes `--inline=no`.

**Comment by yuyichao** [comment]
Ping. What's the status here?

**Comment by wookay** [comment]
@yuyichao hello, there's an error during on test (on OSX, inline=no option for Julia 0.5-dev)
just saw your new pull request (isvalid method).

**Comment by wookay** [comment]
rebase for squashing commits. thanks.

File: deps/build.jl [hunk 16]
```diff
@@ -62,7 +62,8 @@ juliakspec = abspath(spec_name)
 binary_name = @windows? "julia.exe":"julia"
 kernelcmd_array = String[joinpath(JULIA_HOME,("$binary_name")), "-i"]
 ijulia_dir = get(ENV, "IJULIA_DIR", Pkg.dir("IJulia")) # support non-Pkg IJulia installs
-append!(kernelcmd_array, ["-F", joinpath(ijulia_dir,"src","kernel.jl"), "{connection_file}"])
+startupfile = VERSION >= v"0.4-" ? "--startup-file=yes" : "-F"
```
Comment (stevengj) [inline deps/build.jl]: `0.4`, not `0.4-` (which denotes every prerelease of 0.4, from the instant it was forked from 0.3)

File: src/execute_request.jl [hunk 17]
```diff
@@ -72,7 +73,7 @@ end
 
 # return the content of a pyerr message for exception e
 function error_content(e, bt=catch_backtrace(); backtrace_top::Symbol=:execute_request_0x535c5df2, msg::AbstractString="")
-    tb = map(utf8, @compat(split(sprint(show_bt,
+    tb = map(Compat.UTF8String, @compat(split(sprint(show_bt,
```
Comment (stevengj) [inline src/execute_request.jl]: Will this work on 0.3?

File: deps/build.jl (hunk 16 above)
Comment (wookay) [inline deps/build.jl]: Fixed, thanks!

File: src/execute_request.jl (hunk 17 above)
Comment (wookay) [inline src/execute_request.jl]: changed as
```
map(x->convert(Compat.UTF8String, x), ...
```

**Comment by andreasnoack** [comment]
@stevengj Is this ready?

**Comment by stevengj** [comment]
AppVeyor is failing with julia-latest?

**Comment by wookay** [comment]
it's been failed about Conda linking packages.
```
Linking packages ...
Error: Error: post-link failed for: qt-4.8.7-vc9_9
```

**Comment by tkelman** [comment]
cc @Luthaf in case there's a broader issue here

**Comment by wookay** [comment]
Found a workaround for linking qt (on windows julia-0.5-pre)
```
      Pkg.clone(pwd(), \"IJulia\");

      # workaround for linking qt
      using Conda;
      try Conda.add(\"qt\") end;
      postlink = normpath(Pkg.dir(\"Conda\"), \"deps/usr/pkgs/qt-4.8.7-vc9_9/Scripts/.qt-post-link.bat\");
      isfile(postlink) && rm(postlink);

      Pkg.build(\"IJulia\")"
```

**Comment by wookay** [comment]
oh, ready to rebase. waiting reviews. thanks.

File: deps/build.jl [hunk 18]
```diff
@@ -59,10 +59,11 @@ debugdesc = ccall(:jl_is_debugbuild,Cint,())==1 ? "-debug" : ""
 spec_name = "julia-$(VERSION.major).$(VERSION.minor)"*debugdesc
 juliakspec = abspath(spec_name)
 
-binary_name = @windows? "julia.exe":"julia"
+binary_name = @static is_windows() ? "julia.exe" : "julia"
```
Comment (stevengj) [inline deps/build.jl]: No need for `@static` in the build script.

Comment (wookay) [inline deps/build.jl]: Fixed. thanks.

File: appveyor.yml [hunk 19]
```diff
@@ -30,6 +32,13 @@ install:
...
 
 build_script:
-  - C:\projects\julia\bin\julia -e "versioninfo();
-      Pkg.clone(pwd(), \"IJulia\"); Pkg.build(\"IJulia\")"
-#  - C:\projects\julia\bin\julia -e "Pkg.test(\"IJulia\")"
+  - C:\projects\julia\bin\julia -F -e "versioninfo(true);
+      Pkg.clone(pwd(), \"IJulia\");
+      using Conda;
+      try Conda.add(\"qt\") end;
+      postlink = normpath(Pkg.dir(\"Conda\"), \"deps/usr/pkgs/qt-4.8.7-vc9_9/Scripts/.qt-post-link.bat\");
```
Comment (tkelman) [inline appveyor.yml]: this is an odd workaround and seems likely to be fragile?

File: src/stdio.jl [hunk 20]
```diff
@@ -176,7 +176,9 @@ end
 function oslibuv_flush()
     #refs: https://github.com/JuliaLang/IJulia.jl/issues/347#issuecomment-144505862
     #      https://github.com/JuliaLang/IJulia.jl/issues/347#issuecomment-144605024
-    @windows_only ccall(:SwitchToThread, stdcall, Void, ())
+    @compat @static if is_windows()
```
Comment (tkelman) [inline src/stdio.jl]: `@compat` not needed here

File: appveyor.yml (hunk 19 above)
Comment (wookay) [inline appveyor.yml]: right. it's just to pass the appveyor. I don't know exactly how to resolve that correctly.
I thought that this pull request has too much things,
I hope to open another commit/issue/pull request for this responsibility.

**Comment by stevengj** [comment]
Let's just merge this and robustify AppVeyor separately as needed.
